File: packages/contracts/python/butterflylens/contracts/project_run.py
"""Project and run declarations matching the JSON wire contracts."""

from __future__ import annotations

from typing import Literal, TypedDict


PROJECT_SCHEMA_VERSION = "butterflylens-project:v1.0.0"
RUN_SCHEMA_VERSION = "butterflylens-run:v1.0.0"

PROJECT_STATUSES = ("draft", "active", "paused", "archived")
RUN_KINDS = (
    "taxonomy_pack",
    "ala_baseline",
    "reference_bank",
    "flickr_discovery",
    "vision_pipeline",
    "geographic_impact",
    "quality_snapshot",
    "release_export",
    "full_pipeline",
)
RUN_MODES = ("live", "submitted", "replay")
RUN_STATUSES = (
    "queued",
    "leased",
    "running",
    "paused",
    "cancelling",
    "cancelled",
    "succeeded",
    "failed",
)
PIPELINE_STAGE_IDS = (
    "taxonomy",
    "names",
    "ala_baseline",
    "reference_admission",
    "flickr_plan",
    "flickr_fetch",
    "media",
    "yoloe",
    "bioclip",
    "scoring",
    "review",
    "geographic_impact",
    "quality",
    "release",
    "export",
)


class GeographicScope(TypedDict):
    country_code: Literal["AU"]
    boundary_id: str
    boundary_version: str
    boundary_sha256: str
    sensitive_coordinate_policy_version: str


class TaxonScope(TypedDict):
    root_taxon_keys: list[str]
    taxonomy_fingerprint: str


class DiscoveryScope(TypedDict):
    search_plan_fingerprint: str
    public_discovery_claim: Literal[
        "All butterfly candidate images discoverable through the published "
        "ButterflyLens Flickr search plan."
    ]


class ButterflyLensProject(TypedDict):
    schema_version: Literal["butterflylens-project:v1.0.0"]
    project_id: str
    slug: str
    name: str
    description: str
    status: Literal["draft", "active", "paused", "archived"]
    geographic_scope: GeographicScope
    taxon_scope: TaxonScope
    discovery_scope: DiscoveryScope
    data_policy_version: str
    consent_policy_version: str
    created_at: str
    updated_at: str


class RequestedBy(TypedDict):
    actor_type: Literal["system", "account", "operator"]
    actor_id: str | None


class RunEngine(TypedDict):
    repository: str
    commit: str
    interface_version: str
    command: str


class RunStage(TypedDict):
    stage_id: str
    status: Literal[
        "pending",
        "blocked",
        "running",
        "paused",
        "succeeded",
        "failed",
        "cancelled",
        "unavailable",
    ]
    started_at: str | None
    finished_at: str | None
    checkpoint_fingerprint: str | None
    records_processed: int
    records_total: int | None


class RunArtifact(TypedDict):
    artifact_id: str
    kind: str
    object_key: str
    sha256: str
    byte_count: int
    schema_version: str
    semantic_fingerprint: str


class RunError(TypedDict):
    code: str
    message: str
    retryable: bool
    stage_id: str | None


class ButterflyLensRun(TypedDict):
    schema_version: Literal["butterflylens-run:v1.0.0"]
    run_id: str
    project_id: str
    run_kind: str
    mode: Literal["live", "submitted", "replay"]
    status: str
    requested_by: RequestedBy
    requested_at: str
    started_at: str | None
    finished_at: str | None
    updated_at: str
    engine: RunEngine
    input_fingerprints: list[str]
    stages: list[RunStage]
    artifacts: list[RunArtifact]
    error: RunError | None
    revision: int

File: packages/contracts/src/project-run.ts
export const PROJECT_SCHEMA_VERSION = 'butterflylens-project:v1.0.0' as const
export const RUN_SCHEMA_VERSION = 'butterflylens-run:v1.0.0' as const

export const PROJECT_STATUSES = [
  'draft',
  'active',
  'paused',
  'archived',
] as const

export type ProjectStatus = (typeof PROJECT_STATUSES)[number]

export interface ButterflyLensProject {
  readonly schema_version: typeof PROJECT_SCHEMA_VERSION
  readonly project_id: string
  readonly slug: string
  readonly name: string
  readonly description: string
  readonly status: ProjectStatus
  readonly geographic_scope: {
    readonly country_code: 'AU'
    readonly boundary_id: string
    readonly boundary_version: string
    readonly boundary_sha256: string
    readonly sensitive_coordinate_policy_version: string
  }
  readonly taxon_scope: {
    readonly root_taxon_keys: readonly string[]
    readonly taxonomy_fingerprint: string
  }
  readonly discovery_scope: {
    readonly search_plan_fingerprint: string
    readonly public_discovery_claim: 'All butterfly candidate images discoverable through the published ButterflyLens Flickr search plan.'
  }
  readonly data_policy_version: string
  readonly consent_policy_version: string
  readonly created_at: string
  readonly updated_at: string
}

export const RUN_KINDS = [
  'taxonomy_pack',
  'ala_baseline',
  'reference_bank',
  'flickr_discovery',
  'vision_pipeline',
  'geographic_impact',
  'quality_snapshot',
  'release_export',
  'full_pipeline',
] as const

export const RUN_MODES = ['live', 'submitted', 'replay'] as const

export const RUN_STATUSES = [
  'queued',
  'leased',
  'running',
  'paused',
  'cancelling',
  'cancelled',
  'succeeded',
  'failed',
] as const

export const PIPELINE_STAGE_IDS = [
  'taxonomy',
  'names',
  'ala_baseline',
  'reference_admission',
  'flickr_plan',
  'flickr_fetch',
  'media',
  'yoloe',
  'bioclip',
  'scoring',
  'review',
  'geographic_impact',
  'quality',
  'release',
  'export',
] as const

export type RunKind = (typeof RUN_KINDS)[number]
export type RunMode = (typeof RUN_MODES)[number]
export type RunStatus = (typeof RUN_STATUSES)[number]
export type PipelineStageId = (typeof PIPELINE_STAGE_IDS)[number]

export interface RunStage {
  readonly stage_id: PipelineStageId
  readonly status:
    | 'pending'
    | 'blocked'
    | 'running'
    | 'paused'
    | 'succeeded'
    | 'failed'
    | 'cancelled'
    | 'unavailable'
  readonly started_at: string | null
  readonly finished_at: string | null
  readonly checkpoint_fingerprint: string | null
  readonly records_processed: number
  readonly records_total: number | null
}

export interface RunArtifact {
  readonly artifact_id: string
  readonly kind: string
  readonly object_key: string
  readonly sha256: string
  readonly byte_count: number
  readonly schema_version: string
  readonly semantic_fingerprint: string
}

export interface ButterflyLensRun {
  readonly schema_version: typeof RUN_SCHEMA_VERSION
  readonly run_id: string
  readonly project_id: string
  readonly run_kind: RunKind
  readonly mode: RunMode
  readonly status: RunStatus
  readonly requested_by: {
    readonly actor_type: 'system' | 'account' | 'operator'
    readonly actor_id: string | null
  }
  readonly requested_at: string
  readonly started_at: string | null
  readonly finished_at: string | null
  readonly updated_at: string
  readonly engine: {
    readonly repository: string
    readonly commit: string
    readonly interface_version: string
    readonly command: string
  }
  readonly input_fingerprints: readonly string[]
  readonly stages: readonly RunStage[]
  readonly artifacts: readonly RunArtifact[]
  readonly error: {
    readonly code: string
    readonly message: string
    readonly retryable: boolean
    readonly stage_id: string | null
  } | null
  readonly revision: number
}
